src/Pirouette/Term/ToTla.hs

```diff
@@ -816,9 +856,23 @@ trFreeName n args = do
       tlaAppName n args
     DTypeDef _        -> throwError' $ PEOther "trFreeName: Found TypeDef where a term was expected"
 
+-- |Saves a function into the state for defunctionalization and returns its index
+saveDefun :: (MonadPirouette m) => Int -> TLA.AS_Expression -> TlaT m Int
+saveDefun arity fun = do
+  allFuns <- gets tsDefunDefs
+  let thisArityFuns = M.findWithDefault mempty arity allFuns
+  modify' $ \st -> st { tsDefunDefs = M.insert arity (thisArityFuns |> fun) allFuns }
+  pure $ Seq.length thisArityFuns
+
 -- |Constructs a value of a datatype as a TLA expression.
 trConstrApp :: (MonadPirouette m) => Name -> [TLA.AS_Expression] -> TlaT m TLA.AS_Expression
-trConstrApp = tlaAppName
+trConstrApp n args = mapM defun args >>= tlaAppName n
+  where
+    defun arg@(TLA.AS_QuantifierBoundFunction _ funargs _) = do
+      let arity = length funargs
+      curIdx <- saveDefun arity arg
+      pure $ defunLabel arity curIdx
+    defun arg = pure arg
 
```

Comment:
That's interesting! You're doing this at the TLA+ level by saving the body of the function. I do wonder how do we handle free variables though.

Take the following program:
```haskell
main = do
  x <- read <$> getChar
  chars <- concatMap words . lines <$> readFile "whatever"
  map (+ x) chars
```

I'd expect it to be defunctionalized to something like:
```haskell
data LamCharChar = L1 Char

apply :: LamCharChar -> Char -> Char
apply (L1 x) c = c + x

map' f (x:xs) = apply f x : map' f xs

main = do
  x <- read <$> getChar
  chars <- concatMap words . lines <$> readFile "whatever"
  map' (L1 x) chars
```

Note how we keep the `x` around because it is a free variable in the body of the abstraction being defunctionalized.

Reply:
Yep, I wanted to discuss this. Would this be a problem for real-world contracts? Doing this in general requires lambda lifting, which is not the most trivial transformation out there :)

Although, on the other hand, it just occurred to me today that (especially since TLA+ is so untyped) there is a much easier route. We could just store a pair of `("label", [free vars list])` instead of just the `"label"`, which I think should work fine in this case. I'll implement this more generic behaviour.